text/projects/Krzysztof_Pióro/Lyndon_Factorization.tex

```diff
@@ -0,0 +1,146 @@
+    \begin{itemize}
+        \item $v$ nie jest prefiksem $uv$ -- wtedy $vu < uv$, czyli sprzeczność
+        \item $v$ jest prefiksem $uv$ -- tutaj zauważamy, że $u$ jest najmniejsze leksykograficznie spośród prefiksów długości $|u|$ przesunięć cyklicznych słowa $w$.
+        Zatem $vu \leq uv$, czyli sprzeczność.
+    \end{itemize} 
+\end{proof}
+
+\begin{observation}
+    Dla faktoryzacji Lyndona $w = w_1w_2\ldots w_k$ słowo $w_k$ jest minimalnym sufiksem słowa $w$. 
+\end{observation}
+
+\begin{theorem}
+    Dla każdego słowa $w$ istnieje faktoryzacja Lyndona.
+\end{theorem}
+
+\begin{proof}
+    Pojedyncza litera jest słowem \textit{prostym}, zatem możemy zacząć od podziału słowa $w$ na pojedyncze literki.
+    Łatwo zauważyć, że dla dwóch słów prostych $u$, $v$ takich, że $u < v$ słowo $uv$ również jest słowem prostym.
+    Zatem dopóki będą istniały dwa sąsiednie słowa $u$, $v$ w naszej faktoryzacji takie, że $u < v$ to możemy je łączyć w jedno słowo $uv$.
+    Powtarzając tą procedurę otrzymamy faktoryzację Lyndona.
+\end{proof}
+
+\begin{theorem}
+    Dla każdego słowa $w$ istnieje \textbf{unikalna} faktoryzacja Lyndona.
+\end{theorem}
+
+\begin{proof}
+    Z poprzedniego twierdzenia wiemy, że faktoryzacja Lyndona zawsze istnieje. Pozostało nam pokazać jej unikalność.
+    Wykorzystamy do tego celu fakt, że ostatnie słowo z faktoryzacji Lyndona słowa $w$ jest minimalnym sufiksem słowa $w$. 
+    Możemy zatem odciąć minimalny sufiks słowa $w$ i wywołać się indukcyjnie na krótszym słowie.
+\end{proof}
+
+\section{Algorytm Duval'a}
+
+Zacznijmy od wprowadzenia dodatkowej definicji. Słowo nazwiemy \textit{prawie prostym} jeśli jest postaci $w = u^t\bar{u}$, gdzie 
+$\bar{u}$ jest prefiksem słowa $u$ (być może pustym).
+
+Algorytm Duval'a będzie utrzymywał podział słowa wejściowego $w$ na $3$ słowa $w=v_1v_2v_3$ takie, że dla fragmentu $v_1$ faktoryzacja Lyndona jest już znana, 
+a fragment $v_2$ jest słowem \textit{prawie prostym}. 
+
+Algorytm będzie utrzymywał ten podział za pomocą dwóch zmiennych $i, j$. Zmienna $i$ będzie wskazywała na początek słowa $v_2$, a zmienna $j$ będzie wskazywała na początek słowa $v_3$. 
+W każdym kroku algorytmu będziemy próbowali doczepić literę $w[j]$ do słowa $v_2$. W tym celu będziemy porównywali ją z literą słowa $v_2$ wyznaczoną przez zmienną $k$ (taką, że $j - k$ 
+to długość słowa $u$ występującego w \textit{prawie prostym} słowie $v_2 = u^t\bar{u}$). 
+Dokładniej będziemy mieli trzy przypadki:
+\begin{itemize}
+    \item $w[j] = w[k]$: dodanie $w[j]$ do $v_2$ nie narusza założenia, że $v_2$ jest \textit{prawie proste}. Także w tym przypadku po prostu zwiększamy zmienne $j$ oraz $k$.
+    \item $w[j] > w[k]$: słowo $v_2 + w[j]$ staje się proste. W tym przypadku zwiększamy $j$ i ustawiamy $k$ na początek słowa $v_2$.
+    \item $w[j] < w[k]$: słowo $v_2 + w[j]$ przestaje być \textit{prawie proste}. W tym przypadku dla naszego słowa $v_2 = u^t\bar{u}$ dzielimy $u^t$ na $t$ słów \textit{prostych},
```

Comment:
"słowo $v_2 + w[j]$ przestaje być \textit{prawie proste}" - it's not obvious to the reader why